spark/src/main/scala/ai/chronon/spark/Join.scala

```diff
@@ -185,7 +190,64 @@ class Join(joinConf: api.Join,
     coveringSetsPerJoinPart
   }
 
-  override def computeRange(leftDf: DataFrame, leftRange: PartitionRange, bootstrapInfo: BootstrapInfo): DataFrame = {
+  def getAllLeftSideKeyNames(): Seq[String] = {
+    joinConf.getJoinParts.asScala.flatMap { joinPart =>
+      if (joinPart.keyMapping != null) {
+        joinPart.keyMapping.asScala.keys.toSeq
+      } else {
+        joinPart.groupBy.getKeyColumns.asScala
+      }
+    }
+  }
+
+  def injectKeyFilter(leftDf: DataFrame, joinPart: api.JoinPart): Unit = {
+    // Modifies the joinPart to inject the key filter into the
+    val groupByKeyNames = joinPart.groupBy.getKeyColumns.asScala
+
+    // In case the joinPart uses a keymapping
+    val leftSideKeyNames: Map[String, String] = if (joinPart.keyMapping != null) {
+      joinPart.rightToLeft
+    } else {
+      groupByKeyNames.map { k =>
+        (k, k)
+      }.toMap
+    }
+
+    joinPart.groupBy.sources.asScala.foreach { source =>
+      val selectMap = Option(source.rootQuery.getQuerySelects).getOrElse(Map.empty[String, String])
+      val groupByKeyExpressions = groupByKeyNames.map { key =>
+        key -> selectMap.getOrElse(key, key)
+      }.toMap
+
+      groupByKeyExpressions.map{ case (keyName, groupByKeyExpression) =>
+        val leftSideKeyName = leftSideKeyNames.get(keyName).get
+        val values = leftDf.select(leftSideKeyName).collect().map(row => row(0))
```

Comment:
we can run `.collect()` only once to pull all keys locally , instead of once for each key